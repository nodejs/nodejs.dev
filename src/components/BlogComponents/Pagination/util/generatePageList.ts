import { PageModel, PageType } from '../../../../types/pagination';

export function generatePageList(
  pageCount: number,
  currentPage: number,
  showPages: boolean
): PageModel[] {
  const pages: PageModel[] = [];
  if (showPages) {
    for (let i = 1; i <= pageCount; i += 1) {
      const selected = i === currentPage;
      pages.push({
        type: PageType.NUM,
        num: i,
        selected,
      });
    }
  }

  return pages;
}
