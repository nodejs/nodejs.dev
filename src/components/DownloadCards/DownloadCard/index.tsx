import React from 'react';
import './DownloadCard.scss';
import classnames from 'classnames';

interface Props {
  name: string;
  icon: string;
  label: string;
  download: string;
  fileName: string;
  selected: boolean;
  onSelect: (name: string) => void;
}

export default function DownloadCard({
  name,
  icon,
  label,
  download,
  fileName,
  selected,
  onSelect,
}: Props): JSX.Element {
  const handleSelectCard = (): void => onSelect(name);
  const classNames = classnames('download-card', {
    'download-card--active': selected,
  });
  return (
    <li
      className={classNames}
      key={name}
      role="presentation"
      onClick={handleSelectCard}
    >
      <div className="download-card__top">
        <img
          className="download-card__image"
          src={icon}
          alt={`${label} logo`}
        />
        {selected && (
          <a className="download-card__link" href={download}>
            <i className="material-icons">get_app</i>
          </a>
        )}
      </div>
      <p className="download-card__label">{label}</p>
      <p className="download-card__filename">{fileName}</p>
    </li>
  );
}
