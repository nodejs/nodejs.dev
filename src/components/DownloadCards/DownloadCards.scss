.download-cards {
  display: flex;
  align-items: center;
  justify-content: space-between;
  list-style: none;
  padding: 0px;
  margin-block-end: 0px;
  margin-block-start: 0px;
  width: 100%;
  max-width: 615px;
  margin-top: 40px;
}
