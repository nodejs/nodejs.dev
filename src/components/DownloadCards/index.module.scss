.download-cards-wrapper {
  list-style: none;
  width: 100%;

  &:focus {
    outline: none;
  }

  .download-cards {
    align-items: center;
    display: flex;
    justify-content: space-between;
    list-style: none;
    margin: 40px auto 0 auto;
    margin-block-end: 0px;
    margin-block-start: 0px;
    max-width: 615px;
    padding: 0px;
    width: 100%;

    @media (max-width: 460px) {
      flex-direction: column;
      margin-top: 10px;
    }
  }
}
