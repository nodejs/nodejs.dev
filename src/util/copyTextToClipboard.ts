export const copyTextToClipboard = (value: string) => {
  if (typeof window === 'undefined' || typeof navigator === 'undefined') {
    return Promise.resolve(false);
  }

  return navigator.clipboard
    .writeText(value)
    .then(() => true)
    .catch(() => false);
};
