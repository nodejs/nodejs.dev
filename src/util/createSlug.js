function createSlug(title) {
  let slug = title.toLowerCase().trim();
  
  const sets = [
    {to: 'nodejs', from: /node.js/ }, // Replace node.js
    {to: '-and-', from: /&/ }, // Replace &
    {to: '-', from: /(\/|_|,|:|;|\\|\ |\.)/g } // Replace /_,:;\. and whitespace
  ];
  
  sets.forEach(set => {
    slug = slug.replace(set.from, set.to);
  });

  return slug
    .replace(/[^\w\-]+/g, '') // Remove any non word characters
    .replace(/\-\-+/g, '-');  // Replace multiple hyphens
}

module.exports = createSlug;
