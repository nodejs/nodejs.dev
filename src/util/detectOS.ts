export enum UserOS {
  MAC = 'MAC',
  WIN = 'WIN',
  UNIX = 'UNIX',
  LINUX = 'LINUX',
  MOBILE = 'MOBILE',
  UNKNOWN = 'UNKNOWN',
}

export function detectOS(): UserOS {
  let OS = UserOS.UNKNOWN;
  if (typeof window !== 'undefined') {
    switch (typeof navigator !== 'undefined') {
      case navigator.appVersion.includes('Win'):
        OS = UserOS.WIN;
        break;
      case navigator.appVersion.includes('Mac'):
        OS = UserOS.MAC;
        break;
      case navigator.appVersion.includes('X11'):
        OS = UserOS.UNIX;
        break;
      case navigator.appVersion.includes('Linux'):
        OS = UserOS.LINUX;
        break;
      case navigator.appVersion.includes('Mobi'):
        OS = UserOS.MOBILE;
        break;
      default:
        OS = UserOS.UNKNOWN;
        break;
    }
  }
  return OS;
}
