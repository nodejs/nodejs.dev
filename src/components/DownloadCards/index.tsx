import React, { useState, useEffect } from 'react';

import { ReleaseData } from '../../hooks/useReleaseHistory';
import appleLogo from '../../images/logos/apple-logo.svg';
import microsoftLogo from '../../images/logos/microsoft-download-logo.svg';
import sourceCodeIcon from '../../images/logos/source-code-icon.svg';
import { UserOS } from '../../util/detectOS';
import DownloadCard from './DownloadCard';
import './DownloadCards.scss';

interface Props {
  line?: ReleaseData;
  userOS: UserOS;
}

export default function DownloadCards({ line, userOS }: Props): JSX.Element {
  const fileName = line?.version;
  const [selected, setSelected] = useState('');

  useEffect(() => {
    setSelected(
      ![UserOS.WIN, UserOS.MAC, UserOS.MOBILE].includes(userOS)
        ? 'SOURCECODE'
        : userOS
    );
  }, [userOS]);

  const downloadTypes = [
    {
      label: 'Windows Installer',
      icon: microsoftLogo,
      name: UserOS.WIN,
      fileName: `node-${fileName}-x86.msi`,
      download: `https://nodejs.org/dist/${fileName}/node-${fileName}-x86.msi`,
    },
    {
      label: 'MAC Installer',
      icon: appleLogo,
      name: UserOS.MAC,
      fileName: `node-${fileName}.pkg`,
      download: `https://nodejs.org/dist/${fileName}/node-${fileName}.pkg`,
    },
    {
      label: 'Source Code',
      name: 'SOURCECODE',
      icon: sourceCodeIcon,
      fileName: `node-${fileName}.tar.gz`,
      download: `https://nodejs.org/dist/${fileName}/node-${fileName}.tar.gz`,
    },
  ];

  return (
    <ul
      className="download-cards"
      role="tablist"
      tabIndex={0}
      onKeyDown={(e: React.KeyboardEvent): void => {
        const currentIndex = downloadTypes.findIndex(d => d.name === selected);

        let direction = null;
        if (e.key === 'ArrowLeft') {
          direction = 'left';
        }
        if (e.key === 'ArrowRight') {
          direction = 'right';
        }
        if (!direction) return;

        let nextIndex = currentIndex;
        nextIndex += direction === 'left' ? -1 : 1;
        if (nextIndex < 0) {
          nextIndex = downloadTypes.length - 1;
        } else if (nextIndex >= downloadTypes.length) {
          nextIndex = 0;
        }

        const nextItem = downloadTypes[nextIndex].name;
        setSelected(nextItem);
      }}
    >
      {downloadTypes.map(
        (os): JSX.Element => {
          return (
            <DownloadCard
              key={os.name}
              name={os.name}
              icon={os.icon}
              label={os.label}
              download={os.download}
              fileName={os.fileName}
              selected={selected === os.name}
              onSelect={setSelected}
            />
          );
        }
      )}
    </ul>
  );
}
