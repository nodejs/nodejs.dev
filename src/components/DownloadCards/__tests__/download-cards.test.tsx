import React from 'react';
import { fireEvent, render } from '@testing-library/react';

import DownloadCards from '..';
import { UserOS } from '../../../util/detectOS';

describe('DownloadCards component', (): void => {
  it('renders correctly', (): void => {
    const { container } = render(<DownloadCards userOS={UserOS.WIN} />);
    expect(container).toMatchSnapshot();
  });

  it('check click handler on DownloadCards component', (): void => {
    const { getAllByRole } = render(<DownloadCards userOS={UserOS.MAC} />);

    const listElement = getAllByRole('presentation');

    expect(listElement[0]).toHaveClass('download-card');

    fireEvent.click(listElement[0]);

    expect(listElement[0]).toHaveClass('download-card download-card--active');
  });

  it('check left and right arrow click handler on DownloadCards component', (): void => {
    const { getByRole, getAllByRole } = render(
      <DownloadCards userOS={UserOS.MAC} />
    );

    const tabListElement = getByRole('tablist');
    const listElement = getAllByRole('presentation');

    expect(listElement[0]).toHaveClass('download-card');

    /**
     * First left click from default OS of MAC
     */
    fireEvent.keyDown(tabListElement, {
      key: 'ArrowLeft',
      code: 'LeftArrowKey',
    });

    expect(listElement[0]).toHaveClass('download-card download-card--active');
    expect(listElement[1]).toHaveClass('download-card');
    expect(listElement[2]).toHaveClass('download-card');

    /**
     * Left click from WIN
     */
    fireEvent.keyDown(tabListElement, {
      key: 'ArrowLeft',
      code: 'LeftArrowKey',
    });

    expect(listElement[0]).toHaveClass('download-card');
    expect(listElement[1]).toHaveClass('download-card');
    expect(listElement[2]).toHaveClass('download-card download-card--active');

    /**
     * Right click from SOURCECODE
     */

    fireEvent.keyDown(tabListElement, {
      key: 'ArrowRight',
      code: 'RightArrowKey',
    });

    expect(listElement[0]).toHaveClass('download-card download-card--active');
    expect(listElement[1]).toHaveClass('download-card');
    expect(listElement[2]).toHaveClass('download-card');
  });
});
