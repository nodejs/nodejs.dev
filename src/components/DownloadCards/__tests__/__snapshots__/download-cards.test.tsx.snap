// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DownloadCards component renders correctly 1`] = `
<div>
  <ul
    class="download-cards"
    role="tablist"
    tabindex="0"
  >
    <li
      class="download-card download-card--active"
      role="presentation"
    >
      <div
        class="download-card__top"
      >
        <img
          alt="Windows Installer logo"
          class="download-card__image"
          src="test-file-stub"
        />
        <a
          class="download-card__link"
          href="https://nodejs.org/dist/undefined/node-undefined-x86.msi"
        >
          <i
            class="material-icons"
          >
            get_app
          </i>
        </a>
      </div>
      <p
        class="download-card__label"
      >
        Windows Installer
      </p>
      <p
        class="download-card__filename"
      >
        node-undefined-x86.msi
      </p>
    </li>
    <li
      class="download-card"
      role="presentation"
    >
      <div
        class="download-card__top"
      >
        <img
          alt="MAC Installer logo"
          class="download-card__image"
          src="test-file-stub"
        />
      </div>
      <p
        class="download-card__label"
      >
        MAC Installer
      </p>
      <p
        class="download-card__filename"
      >
        node-undefined.pkg
      </p>
    </li>
    <li
      class="download-card"
      role="presentation"
    >
      <div
        class="download-card__top"
      >
        <img
          alt="Source Code logo"
          class="download-card__image"
          src="test-file-stub"
        />
      </div>
      <p
        class="download-card__label"
      >
        Source Code
      </p>
      <p
        class="download-card__filename"
      >
        node-undefined.tar.gz
      </p>
    </li>
  </ul>
</div>
`;
